src/utils.ts

```diff
@@ -0,0 +1,104 @@
+      if (interpolate) {
+        // Bi-linear interpolation.
+        const bottomLeft = image[imageRowIndex][imageColIndex];
+        const bottomRight = imageColIndex + 1 < imageWidth ? image[imageRowIndex][imageColIndex + 1] : bottomLeft;
+        const topLeft = imageRowIndex - 1 >= 0 ? image[imageRowIndex - 1][imageColIndex] : bottomLeft;
+        const topRight = imageRowIndex - 1 >= 0 ?
+          (imageColIndex + 1 < imageWidth ? image[imageRowIndex - 1][imageColIndex + 1] : topLeft) : bottomRight;
+        value = bottomLeft * (1 - imageColAdvance) * (1 - imageRowAdvance) +
+                bottomRight * imageColAdvance * (1 - imageRowAdvance) +
+                topLeft * (1 - imageColAdvance) * imageRowAdvance +
+                topRight * imageColAdvance * imageRowAdvance;
+      }
+      arr.push(value);
+      imageColAdvance += numGridCellsPerImageColPixel;
+      if (imageColAdvance >= 1) {
+        imageColIndex += Math.floor(imageColAdvance);
+        imageColAdvance -= Math.floor(imageColAdvance);
+      }
+    }
+    imageRowAdvance += numGridCellsPerImageRowPixel;
+    if (imageRowAdvance >= 1) {
+      imageRowIndex -= Math.floor(imageRowAdvance);
+      imageRowAdvance -= Math.floor(imageRowAdvance);
+    }
+  }
+  return arr;
+};
+
+// Returns transformed image data.
+export const getImageData = (
+  imgSrc: string,
+  // Function that transfers [r, g, b, a] array into single value
+  transform: (rgba: [number, number, number, number]) => number,
+  // Final callback when data is ready and processed.
+  callback: (imgData: number[][]) => void) => {
+  interface IImageLoadedEvent {
+    target: EventTarget | null;
+  }
+
+  const imageLoaded = (event: IImageLoadedEvent) => {
+    const img = event.target as HTMLImageElement;
+    const canvas = document.createElement("canvas");
+    canvas.width = img.width;
+    canvas.height = img.height;
+    const ctx = canvas.getContext("2d");
+    if (ctx === null) {
+      throw new Error("can't get 2d canvas context");
+    }
+    ctx.drawImage(img, 0, 0, img.width, img.height);
+    const rawData: ImageData = ctx.getImageData(0, 0, img.width, img.height);
+    const data: number[][] = [];
+
+    for (let y = 0; y < rawData.height; y += 1) {
+      const row: number[] = [];
+      data.push(row);
+      for (let x = 0; x < rawData.width * 4; x += 4) {
+        const rIdx = y * (rawData.width * 4) + x;
+        row.push(transform([
+          rawData.data[rIdx],
+          rawData.data[rIdx + 1],
+          rawData.data[rIdx + 2],
+          rawData.data[rIdx + 3]
+        ]));
+      }
+    }
+    callback(data);
+  };
+
+  // Load image first.
+  const imgage = document.createElement("img");
+  imgage.src = imgSrc;
+  if (imgage.complete) {
+    imageLoaded({target: imgage});
+  } else {
+    imgage.addEventListener("load", imageLoaded);
+    imgage.addEventListener("error", () => {
+      throw new Error(`Cannot load image ${imgSrc}`);
+    });
+  }
+};
```

Comment:
This is very nice. Could be separated into it's own little library function if we keep needing to do this.

Reply:
Yeap. Tectonic uses something similar, but it's a bit different though (no transformations and then other library handling `ImageData` directly). Since it's 50 lines and pretty straightforward, I'd leave it unless another project needs that.